clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1085,12 +1085,19 @@ static target getAccessTarget(QualType FieldTy,
       AccTy->getTemplateArgs()[3].getAsIntegral().getExtValue());
 }
 
-// FIXME: Free functions must have void return type, be declared at file scope,
-// outside any namespaces, and with the SYCL_DEVICE attribute. If the
-// SYCL_DEVICE attribute is not specified this function is not entered since the
-// possibility of the function being a free function is ruled out already.
+static bool hasDependentExpr(Expr **Exprs, const size_t ExprsSize) {
+  return std::any_of(Exprs, Exprs + ExprsSize, [](const Expr *E) {
+    return E->isValueDependent() || E->isTypeDependent();
+  });
+}
+
+// FIXME: Free functions must be declared at file scope, outside any namespaces.
 static bool isFreeFunction(SemaSYCL &SemaSYCLRef, const FunctionDecl *FD) {
   for (auto *IRAttr : FD->specific_attrs<SYCLAddIRAttributesFunctionAttr>()) {
+    // Free function properties are all compiletime constants, so skip checking
+    // any attribute values that are not constants.
+    if (hasDependentExpr(IRAttr->args_begin(), IRAttr->args_size()))
```

Comment:
I'm not certain I understand this... dependence has nothing to do with whether something is or isn't constant. The general idea is: Sema checks some non-dependent conditions so it can give early diagnostics before a template is instantiated, but all the dependent conditions eventually become non-dependent once the template is instantiated and that's when the conditions are checked.

So I think this comment needs to be updated, or I'm misunderstanding something about the changes.

Reply:
While searching for other uses of `getAttributeNameValuePairs `I discovered that it is prefaced with the hasDependentExpr check in a couple of other places. That worked in this case too, so I adopted it.

Reply:
Yeah, I'm not saying the code is wrong, I'm saying the comment associated with the code is wrong.